### PR TITLE
refact: implement tokio asynchrony + general cleanup

### PR DESCRIPTION
**Pull request by cartercanedy**

### CODE REVIEW

**Comment by cartercanedy** [comment]
@adamperkowski since you're not in the git blame of this diff and you don't have write access, I can't add you as a reviewer, but I'd appreciate having a second pair of eyes checking over this

**Review by adamperkowski — APPROVED** [review]
Code itself looks okay, although you could
- use something like thiserror, or even anyhow insted of handling errors semi-manually
- create custom types or structs for some return values
- split functions into smaller ones
- write more comments
- move internal logic into `lib.rs` and write documentation comments.
for example:
```diff
-// rawbit is ...
+//! rawbit is ...
```
or
```diff
-// this function ...
+/// this function ...
```

**Pull request merged**
